File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        return dfs(node, new HashMap<>());
    }

    private Node dfs(Node node, HashMap<Node, Node> oldToNewMap){
        if(node == null){
            return null;
        }

        if(oldToNewMap.containsKey(node)){
            return oldToNewMap.get(node);
        }

        Node clone = new Node(node.val);
        oldToNewMap.put(node, clone);

        for(Node neigh: node.neighbors){
            clone.neighbors.add(dfs(neigh, oldToNewMap));
        }

        return clone;
    }
}

File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        return self.dfs(node, {})
    
    def dfs(self, node, oldToNewMap) -> Optional['Node']:
        if node == None:
            return None
        
        if node in oldToNewMap:
            return oldToNewMap[node]
        
        clone = Node(node.val)
        oldToNewMap[node] = clone
        for neigh in node.neighbors:
            clone.neighbors.append(self.dfs(neigh, oldToNewMap))
        
        return clone